File: src/libhalodb/impl/libhalodb.clj
(ns libhalodb.impl.libhalodb
  (:require
   [clj-halodb.core :as halodb])
  (:gen-class
   :methods [^{:static true} [halodbOpen [String] Integer]
             ^{:static true} [halodbClose [] Integer]
             ^{:static true} [halodbSize [] Long]
             ^{:static true} [halodbPut [String String] Integer]
             ^{:static true} [halodbGet [String] String]
             ^{:static true} [halodbDelete [String] Integer]
             ^{:static true} [halodbPauseCompaction [] Integer]
             ^{:static true} [halodbResumeCompaction [] Integer]]))

(set! *warn-on-reflection* true)

(System/setProperty "org.caffinitas.ohc.allocator" "unsafe")

(def options
  (halodb/options
    {:max-file-size (* 1024 1024)
     :sync-write false
     :use-memory-pool true
     :memory-pool-chunk-size (* 1024 (* 16 1024))
     :compaction-job-rate (* 1024 (* 50 1024))
     :fixed-key-size 256}))

(def db
  (atom nil))

(defn open [dbpath]
  (halodb/open dbpath options))

(defn -halodbOpen [path]
  (int
    (try
      (reset! db (open path))
      0
      (catch Exception e
        (println e)
        1))))

(defn -halodbClose []
  (int
    (let [hdb (deref db)]
      (if (halodb/db? hdb)
        (try
          (halodb/close hdb)
          (reset! db nil)
          0
          (catch Exception e
            (println e)
            1))
        -1))))

(defn -halodbSize []
  (let [db (deref db)]
    (if (halodb/db? db)
      (try
        (halodb/size db)
        (catch Exception e
          (println e)
          -2))
      -1)))

(defn -halodbPut [k v]
  (int
    (let [db (deref db)]
      (if (halodb/db? db)
        (try
          (doall
            (halodb/put db {k v}))
          0
          (catch Exception e
            (println e)
            1))
        -1))))

(defn -halodbGet [k]
  (let [db (deref db)]
    (if (halodb/db? db)
      (try
        (halodb/get db k)
        (catch Exception e
          (println e)
          ""))
      "")))

(defn -halodbDelete [k]
  (int
    (let [db (deref db)]
      (if (halodb/db? db)
        (try
          (halodb/delete db k)
          0
          (catch Exception e
            (println e)
            1))
        -1))))

(defn -halodbPauseCompaction []
  (int
    (let [db (deref db)]
      (if (halodb/db? db)
        (try
          (halodb/pause-compaction db)
          0
          (catch Exception e
            (println e)
            1))
        -1))))

(defn -halodbResumeCompaction []
  (int
    (let [db (deref db)]
      (if (halodb/db? db)
        (try
          (halodb/resume-compaction db)
          0
          (catch Exception e
            (println e)
            1))
        -1))))

File: src/libhalodb/impl/LibHaloDB.java
package libhalodb.impl;

import org.graalvm.nativeimage.IsolateThread;
import org.graalvm.nativeimage.c.function.CEntryPoint;
import org.graalvm.nativeimage.c.type.CCharPointer;
import org.graalvm.nativeimage.c.type.CTypeConversion;
import com.oracle.svm.core.c.CConst;

public final class LibHaloDB {
    @CEntryPoint(name = "halodb_open")
    public static int halodbOpen(IsolateThread thread, @CConst CCharPointer s) {
        String path = CTypeConversion.toJavaString(s);
        return libhalodb.halodbOpen(path);
    }

    @CEntryPoint(name = "halodb_close")
    public static int halodbClose(IsolateThread thread) {
        return libhalodb.halodbClose();
    }

    @CEntryPoint(name = "halodb_size")
    public static long halodbSize(IsolateThread thread) {
        return libhalodb.halodbSize();
    }

    @CEntryPoint(name = "halodb_put")
    public static int halodbPut(IsolateThread thread, @CConst CCharPointer ks, @CConst CCharPointer vs) {
        String key = CTypeConversion.toJavaString(ks);
        String val = CTypeConversion.toJavaString(vs);
        return libhalodb.halodbPut(key, val);
    }

    @CEntryPoint(name = "halodb_get")
    public static @CConst CCharPointer halodbGet(IsolateThread thread, @CConst CCharPointer s) {
        String key = CTypeConversion.toJavaString(s);
        String result = libhalodb.halodbGet(key);
        CTypeConversion.CCharPointerHolder holder = CTypeConversion.toCString(result);
        CCharPointer value = holder.get();
        return value;
    }

    @CEntryPoint(name = "halodb_delete")
    public static int halodbDelete(IsolateThread thread, @CConst CCharPointer s) {
        String key = CTypeConversion.toJavaString(s);
        return libhalodb.halodbDelete(key);
    }

    @CEntryPoint(name = "halodb_pause_compaction")
    public static int halodbPauseCompaction(IsolateThread thread) {
        return libhalodb.halodbPauseCompaction();
    }

    @CEntryPoint(name = "halodb_resume_compaction")
    public static int halodbResumeCompaction(IsolateThread thread) {
        return libhalodb.halodbResumeCompaction();
    }
}
